# IT3212 Assignment 1 - Student Graduation Dataset

## 1a. Data exploration

In [1]:
import pandas as pd

df = pd.read_csv("../data/graduation_dataset.csv")

display(df.head())
display(df.describe(include="all"))
display(df.dtypes)

,Marital status,Application mode,Application order,Course,Daytime/evening attendance,Previous qualification,Nacionality,Mother's qualification,Father's qualification,Mother's occupation,...,Curricular units 2nd sem (credited),Curricular units 2nd sem (enrolled),Curricular units 2nd sem (evaluations),Curricular units 2nd sem (approved),Curricular units 2nd sem (grade),Curricular units 2nd sem (without evaluations),Unemployment rate,Inflation rate,GDP,Target
0,1,8,5,2,1,1,1,13,10,6,...,0,0,0,0,0.000000,0,10.8,1.4,1.74,Dropout
1,1,6,1,11,1,1,1,1,3,4,...,0,6,6,6,13.666667,0,13.9,-0.3,0.79,Graduate
2,1,1,5,5,1,1,1,22,27,10,...,0,6,0,0,0.000000,0,10.8,1.4,1.74,Dropout
3,1,8,2,15,1,1,1,23,27,6,...,0,6,10,5,12.400000,0,9.4,-0.8,-3.12,Graduate
4,2,12,1,3,0,1,1,22,28,10,...,0,6,6,6,13.000000,0,13.9,-0.3,0.79,Graduate


,Marital status,Application mode,Application order,Course,Daytime/evening attendance,Previous qualification,Nacionality,Mother's qualification,Father's qualification,Mother's occupation,...,Curricular units 2nd sem (credited),Curricular units 2nd sem (enrolled),Curricular units 2nd sem (evaluations),Curricular units 2nd sem (approved),Curricular units 2nd sem (grade),Curricular units 2nd sem (without evaluations),Unemployment rate,Inflation rate,GDP,Target
count,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,...,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424.000000,4424
unique,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,3
top,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,Graduate
freq,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2209
mean,1.178571,6.886980,1.727848,9.899186,0.890823,2.531420,1.254521,12.322107,16.455244,7.317812,...,0.541817,6.232143,8.063291,4.435805,10.230206,0.150316,11.566139,1.228029,0.001969,NaN
std,0.605747,5.298964,1.313793,4.331792,0.311897,3.963707,1.748447,9.026251,11.044800,3.997828,...,1.918546,2.195951,3.947951,3.014764,5.210808,0.753774,2.663850,1.382711,2.269935,NaN
min,1.000000,1.000000,0.000000,1.000000,0.000000,1.000000,1.000000,1.000000,1.000000,1.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,7.600000,-0.800000,-4.060000,NaN
25%,1.000000,1.000000,1.000000,6.000000,1.000000,1.000000,1.000000,2.000000,3.000000,5.000000,...,0.000000,5.000000,6.000000,2.000000,10.750000,0.000000,9.400000,0.300000,-1.700000,NaN
50%,1.000000,8.000000,1.000000,10.000000,1.000000,1.000000,1.000000,13.000000,14.000000,6.000000,...,0.000000,6.000000,8.000000,5.000000,12.200000,0.000000,11.100000,1.400000,0.320000,NaN
75%,1.000000,12.000000,2.000000,13.000000,1.000000,1.000000,1.000000,22.000000,27.000000,10.000000,...,0.000000,7.000000,10.000000,6.000000,13.333333,0.000000,13.900000,2.600000,1.790000,NaN


Marital status                                      int64
Application mode                                    int64
Application order                                   int64
Course                                              int64
Daytime/evening attendance                          int64
Previous qualification                              int64
Nacionality                                         int64
Mother's qualification                              int64
Father's qualification                              int64
Mother's occupation                                 int64
Father's occupation                                 int64
Displaced                                           int64
Educational special needs                           int64
Debtor                                              int64
Tuition fees up to date                             int64
Gender                                              int64
Scholarship holder                                  int64
Age at enrollm

## 1b. Missing values, categorical values and outliers

In [2]:
# Missing values in each column
display(df.isna().sum())

Marital status                                    0
Application mode                                  0
Application order                                 0
Course                                            0
Daytime/evening attendance                        0
Previous qualification                            0
Nacionality                                       0
Mother's qualification                            0
Father's qualification                            0
Mother's occupation                               0
Father's occupation                               0
Displaced                                         0
Educational special needs                         0
Debtor                                            0
Tuition fees up to date                           0
Gender                                            0
Scholarship holder                                0
Age at enrollment                                 0
International                                     0
Curricular u

In [3]:
# These columns contain categories, even though most are stored as numbers
categorical_columns = [
    "Marital status", "Application mode", "Course",
    "Daytime/evening attendance", "Previous qualification",
    "Nacionality", "Mother's qualification",
    "Father's qualification", "Mother's occupation",
    "Father's occupation", "Displaced",
    "Educational special needs", "Debtor",
    "Tuition fees up to date", "Gender",
    "Scholarship holder", "International", "Target"
]

for column in categorical_columns:
    values = sorted(df[column].unique().tolist())
    print(f"{column}: {values}")

Marital status: [1, 2, 3, 4, 5, 6]
Application mode: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18]
Course: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17]
Daytime/evening attendance: [0, 1]
Previous qualification: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17]
Nacionality: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21]
Mother's qualification: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29]
Father's qualification: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34]
Mother's occupation: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32]
Father's occupation: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 4

## 2. Handling missing values

The dataset contains no missing values. Therefore, no imputation or removal is needed, and all 4,424 rows are retained. Zero values are not treated as missing because they may be valid observations.

## 3a. Detecting outliers with the IQR method

Values below $Q1 - 1.5 \times IQR$ or above $Q3 + 1.5 \times IQR$ are identified as potential outliers. IQR was chosen because several numerical variables are skewed and contain many zero values. Unlike Z-score, IQR does not assume an approximately normal distribution and is less affected by extreme values.

In [4]:
# Detect outliers in numerical columns using the IQR method
numerical_columns = [
    column for column in df.columns if column not in categorical_columns
]

Q1 = df[numerical_columns].quantile(0.25)
Q3 = df[numerical_columns].quantile(0.75)
IQR = Q3 - Q1

lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR
outliers = (df[numerical_columns] < lower_bound) | (df[numerical_columns] > upper_bound)

outlier_summary = pd.DataFrame({
    "Lower bound": lower_bound,
    "Upper bound": upper_bound,
    "Number of outliers": outliers.sum()
})

display(outlier_summary[outlier_summary["Number of outliers"] > 0])

,Lower bound,Upper bound,Number of outliers
Application order,-0.500,3.500000,541
Age at enrollment,10.000,34.000000,441
Curricular units 1st sem (credited),0.000,0.000000,577
Curricular units 1st sem (enrolled),2.000,10.000000,424
Curricular units 1st sem (evaluations),0.000,16.000000,158
Curricular units 1st sem (approved),-1.500,10.500000,180
Curricular units 1st sem (grade),7.400,17.000000,726
Curricular units 1st sem (without evaluations),0.000,0.000000,294
Curricular units 2nd sem (credited),0.000,0.000000,530
Curricular units 2nd sem (enrolled),2.000,10.000000,369


## 3b. Handling outliers

The potential outliers are retained. The most extreme values were checked: the maximum age is 70, the maximum application order is 9, the maximum grade is 18.875 on a 0-20 scale, and the maximum number of first-semester evaluations is 45. These values are unusual but possible. No student has more approved units than evaluations, and no student with approved units has a grade of zero. Several columns also contain many valid zero values, causing the IQR method to classify any higher value as an outlier. There is therefore no clear evidence of recording errors, and removing, capping or transforming these observations could remove useful information and distort the dataset.

## 4a. Encoding categorical data

Categorical features with more than two categories are one-hot encoded because their numerical codes represent labels, not quantities or order. Binary features are already represented by 0 and 1 and are kept unchanged. The target is label encoded because classification models require numerical class labels.

In [5]:
X = df.drop(columns="Target")
y = df["Target"].map({"Dropout": 0, "Enrolled": 1, "Graduate": 2})

categorical_features = [
    column for column in categorical_columns if column != "Target"
]
columns_to_encode = [
    column for column in categorical_features if df[column].nunique() > 2
]

X_encoded = pd.get_dummies(X, columns=columns_to_encode, dtype=int)

print("Encoded feature shape:", X_encoded.shape)
display(X_encoded.head())
display(y.head())

Encoded feature shape: (4424, 245)


,Application order,Daytime/evening attendance,Displaced,Educational special needs,Debtor,Tuition fees up to date,Gender,Scholarship holder,Age at enrollment,International,...,Father's occupation_37,Father's occupation_38,Father's occupation_39,Father's occupation_40,Father's occupation_41,Father's occupation_42,Father's occupation_43,Father's occupation_44,Father's occupation_45,Father's occupation_46
0,5,1,1,0,0,1,1,0,20,0,...,0,0,0,0,0,0,0,0,0,0
1,1,1,1,0,0,0,1,0,19,0,...,0,0,0,0,0,0,0,0,0,0
2,5,1,1,0,0,0,1,0,19,0,...,0,0,0,0,0,0,0,0,0,0
3,2,1,1,0,0,1,0,0,20,0,...,0,0,0,0,0,0,0,0,0,0
4,1,0,0,0,0,1,0,0,45,0,...,0,0,0,0,0,0,0,0,0,0


0    0
1    2
2    0
3    2
4    2
Name: Target, dtype: int64

## 4b. Feature scaling

The numerical features are standardized to have a mean of 0 and a standard deviation of 1. Standardization is used because the features have different units and ranges. It prevents features with large values from having an unfair influence on models that use distances or gradient-based optimization. Binary and one-hot encoded features are not scaled. Although data splitting is discussed in section 5, the split is performed here before scaling because StandardScaler learns the mean and standard deviation from the data. Fitting it only on the training data prevents information from the test data from leaking into the training process.

In [6]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

X_train, X_test, y_train, y_test = train_test_split(
    X_encoded, y, test_size=0.2, random_state=42, stratify=y
)

scaler = StandardScaler()
X_train_scaled = X_train.copy()
X_test_scaled = X_test.copy()

X_train_scaled[numerical_columns] = scaler.fit_transform(
    X_train[numerical_columns]
)
X_test_scaled[numerical_columns] = scaler.transform(
    X_test[numerical_columns]
)

display(X_train_scaled[numerical_columns].head())

,Application order,Age at enrollment,Curricular units 1st sem (credited),Curricular units 1st sem (enrolled),Curricular units 1st sem (evaluations),Curricular units 1st sem (approved),Curricular units 1st sem (grade),Curricular units 1st sem (without evaluations),Curricular units 2nd sem (credited),Curricular units 2nd sem (enrolled),Curricular units 2nd sem (evaluations),Curricular units 2nd sem (approved),Curricular units 2nd sem (grade),Curricular units 2nd sem (without evaluations),Unemployment rate,Inflation rate,GDP
2283,-0.556229,-0.295828,3.481269,1.898375,1.363422,1.711901,0.364851,-0.210957,2.836043,1.700121,1.250659,1.838465,0.304781,-0.195530,0.872138,-1.103473,0.339595
3874,-0.556229,2.334335,-0.301418,-0.507336,-0.786079,-1.519082,-2.181071,1.297276,-0.283546,-0.552829,-0.765664,-1.464407,-1.949336,6.822546,0.421006,1.795972,-0.756629
2281,-0.556229,-0.295828,-0.301418,-0.106384,1.124589,-0.226689,0.231395,-0.210957,-0.283546,-0.102239,1.754739,-0.143258,0.056446,-0.195530,1.473648,1.143597,-1.795621
817,-0.556229,-0.690353,-0.301418,-0.106384,-0.547245,0.419507,0.796015,-0.210957,-0.283546,-0.102239,-0.513624,0.517316,0.502175,-0.195530,0.308223,-0.523584,0.779846
404,-0.556229,-0.558845,-0.301418,0.695519,-0.069578,0.742606,0.843531,-0.210957,-0.283546,0.798941,-0.009543,0.847603,0.864762,-0.195530,-0.293287,0.128791,0.757833


## 5a. Training and testing split

The dataset is split into 80% training data and 20% testing data. Stratification keeps approximately the same distribution of target classes in both sets. A fixed random state makes the split reproducible.

In [7]:
print("Training rows:", len(X_train_scaled))
print("Testing rows:", len(X_test_scaled))

Training rows: 3539
Testing rows: 885


## 5b. Importance of data splitting

The model learns from the training data, while the testing data is kept unseen for final evaluation. This gives a more realistic measure of how the model performs on new data and helps reveal overfitting. The split is also performed before fitting the scaler, preventing information from the testing data from leaking into the training process.

## 6. Principal Component Analysis (PCA)

PCA is applied to retain 95% of the variance while reducing the number of features. It is fitted only on the training data to prevent data leakage.

In [ ]:
from sklearn.decomposition import PCA

pca = PCA(n_components=0.95)
X_train_pca = pca.fit_transform(X_train_scaled)
X_test_pca = pca.transform(X_test_scaled)

print("Original number of features:", X_train_scaled.shape[1])
print("Number of principal components:", pca.n_components_)
print("Explained variance:", pca.explained_variance_ratio_.sum().round(4))

PCA reduces the dataset from 245 features to 51 principal components while retaining 95.22% of the variance. This can reduce computation time, redundancy and noise. However, 4.78% of the variance is lost, and the principal components are more difficult to interpret than the original features. Whether PCA improves prediction must be evaluated by comparing models with and without PCA.